### ISSUE TITLE
New branch policy

### ISSUE THREAD

**Issue by UlfNorell** [agda/agda#3104]
After some discussion at the Agda meeting we made came to the conclusion that given how we use the `stable-2.5` and `master` branches it would make more sense to call them `master` and `experimental`.

Furthermore, @Saizan expressed a desire to include cubical in the next release (meaning that the next release would be `2.6.0`). Thus, the natural thing to do would be to "freeze" `stable-2.5` and continue all development on `master`. The `experimental` branch can be created right away (maybe best), or optionally wait until we actually have some divergence.

Things to do:
- [x] Update hacking instructions
- [x] Update Travis jobs
- [x] Notify the `agda-dev` mailing list
- [x] Update [Wiki](http://wiki.portal.chalmers.se/agda/pmwiki.php?n=Main.Download)
- [x] Figure out more things that need to be done
- [x] Do those things

**Comment by asr** [comment]
If I understood correctly the propose is

`master`: the (stable) branch to be use for releasing
`experimental`: the branch to be used for new features

If  so, why continue *all* development on `master`? That is, who is going to test the new features in `experimental`?

**Comment by UlfNorell** [comment]
No I mean we should keep working as before but with the new names. For now though, there aren't any new features on `experimental`.

**Comment by asr** [comment]
I see. Thanks!

**Comment by andreasabel** [comment]
Why would we create the experimental branch right away?  We would then have to do pointless merges all the time...

**Comment by UlfNorell** [comment]
They would be fast-forwards so they wouldn't really be in the way. The point would be that we have all the infrastructure and instructions set up for two branches. I'm a bit reluctant to change all of that only to have to add it back once we decide to create `experimental`.

**Comment by andreasabel** [comment]
How can they be fast-forwards if they differ (e.g. in the `Makefile`)?  It appears this would be constant merging into experimental, compilation again, testsuite again...

How would you name the 2 agda versions? `agda` and `agda-experimental`?  Actually, I prefer `agda-X.Y.Z`.
What version would `experimental` be? 2.7.0?
It seems like all the work you have to put in to set up `experimental` can be deferred until it matters...

**Comment by UlfNorell** [comment]
I was thinking they would be exactly the same until we actually need them to diverge. If you absolutely need it to have a different version number then we should wait to create it.

What I'm saying is that some of the work to set it up we have already done (for instance, Travis jobs and hacking instructions) and this would have to be *undone* if we don't have an experimental branch. Maybe that's not a big deal. @asr?

**Comment by UlfNorell** [comment]
@asr do you have any input on this?

The current proposal is to switch over the development to `master` and not add `experimental` until we need it.

**Comment by asr** [comment]
We are using the current work-flow (i.e. `master` and `stable-2.5`) because we need/want that new features:

1. don't block the release of a new version (e.g. the last release didn't include cubical)
2. be tested for many developers as possible

Now, if we want to release from `master` (good!) and keeping 1. and 2., I suggest:

A. frozen `stable-2.5`
B. create `stable-2.6` from the current `master`
C. bump version in `master` to 2.7
D. continue using `master` and `stable-2.6` as usual

**Comment by UlfNorell** [comment]
There are two things we're talking about here:

I wanted to rename the two branches so instead of `stable-2.6` and `master` we have `master` and `experimental` (or `experimental-2.7` or some such), to fix the problem with PRs and the fact that we really are using the stable branch as a master branch.

The other question is whether we wanted to hold off on creating the experimental branch until we actually have something to put there. This would save some Travis time and unnecessary merging until we actually need it.

**Comment by asr** [comment]
> I wanted to rename the two branches so instead of `stable-2.6` and `master` we have `master` and `experimental` (or `experimental-2.7` or some such), to fix the problem with PRs and the fact that we really are using the stable branch as a master branch.

Let's suppose we have `master` and `experimental` working. Who will be the responsable by merging the new commits pushed in `master` into `experimental`? The author of the commit in `master`? The guys working in `experimental`?

**Comment by UlfNorell** [comment]
Same as before: the person pushing to `master`. The (first) proposed change is *only* to rename the branches.

Regarding experimental, I think the most sensible thing to do would be to set it up right away, but not start the actual merge-every-commit-from-master until we have something interesting on experimental.

**Comment by asr** [comment]
TL;DR: make the renaming without setting `experimental` up.

I don't see any *strong* reason for setting `experimental` up right now, that is, I propose to 
set it up when needed.

In relation to the renaming:

1. So how the names `master` and `experimental` have different implicit meaning. I can imagine the following rude conversation:

    someone: please merge your commit to `experimental`

    author of commit in `master`: I don't care the feature in `experimental`, so I'm not going to merge my commit.


2.  I guess we could/should end up with the following situation/branches:

    `master`

    `experimental-feature-1`

    `experimental-feature-2`

    `experimental-feature-n`

    
    I agree with the above situation. Should the person pushing to `master` be the responsable of merging the changes into every `experimental` branch? I don't think so.

**Comment by UlfNorell** [comment]
> author of commit in master: I don't care the feature in experimental, so I'm not going to merge my commit.

This would be equivalent to saying I don't care about the next major release of Agda, which would be a strange thing to say. If the person pushing to `master` is not responsible for merging into `experimental` it puts an undue burden on the people that care about developing and maintaining the new features.

We should (and do) certainly have feature branches, but having a dedicated `experimental` branch gives us a place to test the interaction between new features and gives people interested in the bleeding edge a place to go. In your proposal there would be no way to use both `feature-1` and `feature-2`.

**Comment by WolframKahl** [comment]
> We should (and do) certainly have feature branches, but having a dedicated `experimental` branch gives us a place to test the interaction between new features and gives people interested in the bleeding edge a place to go.

Perhaps it should then have a less `experimental` name, and instead something like `next-master`, `next-stable`, `future`, ...

**Comment by UlfNorell** [comment]
Sure, I suggested `experimental` because the standard library has an experimental branch, but I'd be fine with a different name. My main desire was to have the main branch be called `master` instead of `stable-2.X`.

**Comment by jespercockx** [comment]
I support calling the new branch `future`. There should be clear instructions that the policy is to always merge `master` into `future`. Without such a policy, developing new major features such as cubical-agda would be as good as impossible.

**Comment by asr** [comment]
> I support calling the new branch `future`. There should be clear instructions that the policy is to always merge `master` into `future`

+1 in both, the branch name and the policy.

**Comment by UlfNorell** [comment]
I'm starting the process. Consider `stable-2.5` frozen.

**Comment by UlfNorell** [comment]
Completed. Happy hacking on master! 🎈

**Comment by andreasabel** [comment]
Patch https://github.com/agda/agda/commit/18743484d0a0f01e068287fe50a465d8e0044699 triggered #3154.
